### ISSUE TITLE
Server Host Key Algorithm Mismatch

### ISSUE THREAD

**Issue by porter707** [phpseclib/phpseclib#1331]
Hi,

We still appear to be getting the Server Host Key Algorithm Mismatch error with a clients site on 1.0.13

It works with 1.0.11

I've done some logging around the area of the error:

```
$this->server_host_key_algorithms:
[14-Jan-2019 09:45:56 UTC] Array
(
    [0] => ecdsa-sha2-nistp256
    [1] => ecdsa-sha2-nistp384
    [2] => ecdsa-sha2-nistp521
    [3] => ssh-ed25519
    [4] => rsa-sha2-512
    [5] => rsa-sha2-256
    [6] => ssh-rsa
    [7] => ssh-dss
)

version 1.0.13

$server_host_key_algorithms:
[14-Jan-2019 09:18:35 UTC] Array
(
    [0] => rsa-sha2-256
    [1] => rsa-sha2-512
    [2] => ssh-rsa
    [3] => ssh-dss
)

[14-Jan-2019 09:18:35 UTC] Public key format: ssh-rsa expected key format: ssh-rsa signature format: ssh-rsa server host key algorithm: rsa-sha2-256

version 1.0.12

$server_host_key_algorithms:
[14-Jan-2019 09:24:15 UTC] Array
(
    [0] => rsa-sha2-256
    [1] => rsa-sha2-512
    [2] => ssh-rsa
    [3] => ssh-dss
)

[14-Jan-2019 09:24:15 UTC] Public key format: ssh-rsa expected key format: ssh-rsa signature format: ssh-rsa server host key algorithm: rsa-sha2-256

version 1.0.11

$server_host_key_algorithms:
[14-Jan-2019 09:28:33 UTC] Array
(
    [0] => ssh-rsa
    [1] => ssh-dss
)
// the if statement is different on 1.0.11
[14-Jan-2019 09:28:33 UTC] Public key format: ssh-rsa server host key algorithm: ssh-rsa signature format: ssh-rsa server host key algorithm: ssh-rsa
```

So it appears _array_intersect_first() is selecting the wrong value? which causes the second half of this if statement to trigger:

```
if ($public_key_format != $expected_key_format || $this->signature_format != $server_host_key_algorithm) {
    user_error('Server Host Key Algorithm Mismatch');
    return $this->_disconnect(NET_SSH2_DISCONNECT_KEY_EXCHANGE_FAILED);
}
```

**Comment by terrafrost** [comment]
> So it appears _array_intersect_first() is selecting the wrong value?

No. `$server_host_key_algorithm` should be `rsa-sha2-256` based on the `print_r` output you've provided. Quoting [RFC4253](https://tools.ietf.org/html/rfc4253#page-18),

```
The first algorithm on the client's name-list
that satisfies the requirements and is also supported by the
server MUST be chosen.  If there is no such algorithm, both
sides MUST disconnect.
```

Maybe `$this->signature_format` is `ssh-rsa` and maybe that's where the mismatch is idk. What'd really help are the SSH logs, which you can get by doing `define('NET_SSH2_LOGGING', 2);` at the top and, later, `echo $ssh->getLog();` after you encounter the error.

Of course, **if** `$this->signature_format` were `ssh-rsa` then the server would be out of compliance with [RFC8332](https://tools.ietf.org/html/rfc8332#section-3). Quoting it,

```
   The resulting signature is encoded as follows:

   string   "rsa-sha2-256" / "rsa-sha2-512"
   string    rsa_signature_blob

   The value for 'rsa_signature_blob' is encoded as a string that
   contains an octet string S (which is the output of RSASSA-PKCS1-v1_5)
   and that has the same length (in octets) as the RSA modulus.  When S
   contains leading zeros, there exist signers that will send a shorter
   encoding of S that omits them.  A verifier MAY accept shorter
   encodings of S with one or more leading zeros omitted. 
```
I mean, I suppose it's possible the server **is** out of compliance. SSH servers sometime do have bugs after all. In that scenario it might be prudent for me to add a special if statement for whatever SSH server you're using. Here's an example of where something similar is already done:

```php
        // some SSH servers have buggy implementations of some of the above algorithms
        switch (true) {
            case $this->server_identifier == 'SSH-2.0-SSHD':
            case substr($this->server_identifier, 0, 13) == 'SSH-2.0-DLINK':
                $mac_algorithms = array_values(array_diff(
                    $mac_algorithms,
                    array('hmac-sha1-96', 'hmac-md5-96')
                ));
        }
```

**Comment by terrafrost** [comment]
And as for `$public_key_format != $expected_key_format`...

Per the `switch` right above that line `$expected_key_format` can only be `ssh-dss` or `ssh-rsa`:

```php
        switch ($server_host_key_algorithm) {
            case 'ssh-dss':
                $expected_key_format = 'ssh-dss';
                break;
            //case 'rsa-sha2-256':
            //case 'rsa-sha2-512':
            //case 'ssh-rsa':
            default:
                $expected_key_format = 'ssh-rsa';
        }
```

As for `$public_key_format`...  quoting [RFC8332](https://tools.ietf.org/html/rfc8332#section-3):

```
   Since RSA keys are not dependent on the choice of hash function, the
   new public key algorithms reuse the "ssh-rsa" public key format as
   defined in [RFC4253]:

   string    "ssh-rsa"
   mpint     e
   mpint     n

   All aspects of the "ssh-rsa" format are kept, including the encoded
   string "ssh-rsa".  This allows existing RSA keys to be used with the
   new public key algorithms, without requiring re-encoding or affecting
   already trusted key fingerprints.
```

So that should be `ssh-rsa` as well. Unless, of course, you have a buggy server.

**Comment by porter707** [comment]
Thanks for the response.

Let me clear some things up.

So I added my own error_log() before the if statement in SSH2.php on line 1815 e.g this one:
```
if ($public_key_format != $expected_key_format || $this->signature_format != $server_host_key_algorithm) {
    user_error('Server Host Key Algorithm Mismatch');
    return $this->_disconnect(NET_SSH2_DISCONNECT_KEY_EXCHANGE_FAILED);
}
```

The above logs for version 1.0.13 and 1.0.12 show that when I log each variable in that if statement I get the following:
```
Public key format: ssh-rsa 
Expected key format: ssh-rsa 
Signature format: ssh-rsa 
Server host key algorithm: rsa-sha2-256
```
So the second part of the if statement is true with these values:
```
$this->signature_format != $server_host_key_algorithm
ssh-rsa != rsa-sh2-256
```

Which results in the server host key algorithm mismatch error

Now in version 1.0.11 the if statement is different
```
if ($public_key_format != $server_host_key_algorithm || $this->signature_format != $server_host_key_algorithm) {
    user_error('Server Host Key Algorithm Mismatch');
    return $this->_disconnect(NET_SSH2_DISCONNECT_KEY_EXCHANGE_FAILED);
}
```

When I error log the variables in that I get this:
```
Public key format: ssh-rsa 
Server host key algorithm: ssh-rsa 
Signature format: ssh-rsa 
Server host key algorithm: ssh-rsa
```

Note that $server_host_key_algorithm is on both sides of the if which is why I was echo'ing it out twice in the last part of my error log dump

Hopefully that clears my message up? 

Maybe it is a buggy server all our clients that have this error appear to be with the same hosting.

Anyway I'll get you the SSH logs and report back here.

**Comment by terrafrost** [comment]
So yah I do contend that the response is malformed. From [RFC4253](https://tools.ietf.org/html/rfc4253#page-22):

      byte      SSH_MSG_KEXDH_REPLY
      string    server public host key and certificates (K_S)
      mpint     f
      string    signature of H

Your response contains ssh-rsa twice:

```
<- NET_SSH2_MSG_KEXDH_REPLY (since last: 0.0559, network: 0.0558s)
00000000  00:00:00:95:00:00:00:07:73:73:68:2d:72:73:61:00  ........ssh-rsa.
00000010  00:00:01:23:00:00:00:81:00:db:32:59:c7:a5:64:48  ...#......2Y..dH
00000020  90:b5:e1:22:0b:aa:fc:cb:89:1d:ef:f1:e4:b4:0f:b6  ..."............
00000030  6c:ec:2e:fb:ef:5c:83:c1:af:74:51:6f:09:92:12:b6  l....\...tQo....
00000040  dd:d9:3f:1b:c4:7b:e2:47:5c:59:81:0e:9f:79:02:4a  ..?..{.G\Y...y.J
00000050  bf:f6:af:3f:94:b5:84:e1:65:e5:94:6b:91:7a:05:c4  ...?....e..k.z..
00000060  ff:5f:91:b0:87:54:40:97:0e:44:4b:19:d1:af:e8:62  ._...T@..DK....b
00000070  66:16:0b:23:b0:62:bd:52:28:e4:4c:de:70:8b:34:28  f..#.b.R(.L.p.4(
00000080  4a:23:31:15:24:5c:8f:82:94:72:1c:c9:70:58:ea:0a  J#1.$\...r..pX..
00000090  df:95:01:60:8d:9a:61:14:cb:00:00:00:80:71:bc:bd  ...`..a......q..
000000a0  dd:02:2d:7f:6a:68:f6:16:07:cf:02:98:24:12:fc:0b  ..-.jh......$...
000000b0  da:02:4f:cc:13:13:c1:6a:ca:7a:a9:de:06:8a:e3:4c  ..O....j.z.....L
000000c0  82:37:83:58:4f:54:ad:f9:7e:1a:82:d4:93:08:bb:cf  .7.XOT..~.......
000000d0  74:5e:50:68:3c:95:b4:32:40:b1:1f:a3:22:2e:fe:a6  t^Ph...2@..."...
000000e0  2c:e9:22:c1:72:14:6a:1c:fb:45:55:e4:be:1a:2d:53  ,.".r.j..EU...-S
000000f0  46:94:2f:2d:c8:42:e5:6d:cc:fa:78:76:b5:56:8d:05  F./-.B.m..xv.V..
00000100  03:39:68:a6:6f:a2:57:64:a3:02:01:68:2f:85:0e:fd  .9h.o.Wd...h/...
00000110  46:d7:d0:f0:ec:86:ca:2b:e4:c2:07:a5:55:00:00:00  F......+....U...
00000120  8f:00:00:00:07:73:73:68:2d:72:73:61:00:00:00:80  .....ssh-rsa....
00000130  31:98:e4:cb:45:63:2b:4d:dd:aa:d8:ce:c2:7f:10:8a  1...Ec+M........
00000140  8b:c3:29:14:51:2a:e2:86:21:ac:75:00:94:b4:b9:03  ..).Q*..!.u.....
00000150  84:23:90:32:37:36:7d:4e:ab:8f:be:7b:eb:f5:ea:9b  .#.276}N...{....
00000160  fa:68:b7:1c:81:74:f5:a8:ec:85:44:75:b2:ba:1b:4d  .h...t....Du...M
00000170  78:1b:c4:fd:18:1e:8b:7c:02:73:69:98:ce:a0:fc:34  x......|.si....4
00000180  df:07:26:0c:4f:25:1c:38:52:16:eb:0b:bb:45:d7:81  ..&.O%.8R....E..
00000190  11:63:40:1b:1b:ab:70:68:69:4c:ee:cf:8f:18:c8:be  .c@...phiL......
000001a0  07:27:42:03:66:b1:89:e4:18:f0:84:eb:2a:27:8d:04  .'B.f.......*'..
```

Here's the response from a digitalocean droplet that I just tried to connect to via SSH:

```
<- NET_SSH2_MSG_KEXDH_REPLY (since last: 0.1584, network: 0.1583s)
00000000  00:00:01:17:00:00:00:07:73:73:68:2d:72:73:61:00  ........ssh-rsa.
00000010  00:00:03:01:00:01:00:00:01:01:00:aa:2a:e2:5c:5e  ............*.\^
00000020  4d:a3:4f:e6:f2:d2:b9:8a:2d:65:3e:41:b4:95:10:6f  M.O.....-e>A...o
00000030  a3:76:da:7a:f2:2c:7e:51:06:a2:c7:e2:ec:a4:4c:cb  .v.z.,~Q......L.
00000040  79:e0:bf:b7:31:3d:0a:65:c4:f5:69:87:02:30:28:4d  y...1=.e..i..0(M
00000050  18:03:7f:30:9d:48:fb:d2:36:63:8f:57:ac:39:8d:74  ...0.H..6c.W.9.t
00000060  cd:1d:4c:82:e1:a7:e7:3f:d5:23:1e:80:b4:72:7a:b8  ..L....?.#...rz.
00000070  89:f8:5b:4a:1a:cc:5f:85:03:8f:08:4e:bc:8e:a3:aa  ..[J.._....N....
00000080  35:b4:eb:a2:50:39:40:40:79:6d:91:07:68:0c:25:bf  5...P9@@ym..h.%.
00000090  66:c8:bc:a0:ab:47:df:0e:6c:1b:50:d4:a2:46:e7:ec  f....G..l.P..F..
000000a0  91:57:15:89:f8:6b:14:f5:6a:a5:a5:ee:31:cb:f9:ab  .W...k..j...1...
000000b0  7e:0e:f7:84:15:9a:b9:1e:a6:23:d7:6f:7f:2c:44:53  ~........#.o.,DS
000000c0  70:e5:cd:ad:c0:7e:86:4f:22:6d:bb:fd:1f:f2:c7:b5  p....~.O"m......
000000d0  0a:74:e8:07:75:50:8e:32:e2:7d:70:f0:0f:4f:2e:ee  .t..uP.2.}p..O..
000000e0  07:f3:89:d7:30:f3:b0:98:8f:c7:ae:a6:f1:85:d0:48  ....0..........H
000000f0  f9:c1:52:c4:f4:7d:84:d9:2f:05:4e:f0:d7:b6:79:31  ..R..}../.N...y1
00000100  69:8d:09:ad:fc:61:74:07:fd:0c:12:b5:41:1b:f8:97  i....at.....A...
00000110  9b:b0:69:42:36:18:f2:5a:e1:30:a1:00:00:01:00:42  ..iB6..Z.0.....B
00000120  a6:c8:dc:9a:63:15:13:14:a2:8d:45:48:61:4c:26:b9  ....c.....EHaL&.
00000130  96:62:bb:34:6b:b7:37:c0:29:dc:72:06:8d:e5:53:6f  .b.4k.7.).r...So
00000140  a4:ea:b5:df:21:99:0f:6a:e4:bd:a1:f1:ed:3d:2e:ef  ....!..j.....=..
00000150  a3:21:bb:74:36:4d:6a:b9:c9:ce:cc:db:93:0b:af:6d  .!.t6Mj........m
00000160  f5:9c:0b:16:3e:cd:4d:42:67:de:d3:91:14:c5:ac:17  ....>.MBg.......
00000170  e9:cc:52:8d:6e:a1:16:4f:f5:19:4e:b1:46:9f:c1:af  ..R.n..O..N.F...
00000180  1e:bf:b8:0d:5e:ea:31:ef:9f:84:aa:c2:5c:4b:35:5d  ....^.1.....\K5]
00000190  c9:aa:64:e4:77:69:dc:c7:78:73:e8:c0:7b:77:1f:10  ..d.wi..xs..{w..
000001a0  33:79:1d:56:73:c8:bf:34:cb:c1:6e:8f:dd:16:c0:d3  3y.Vs..4..n.....
000001b0  fd:25:36:c0:4d:a7:ec:b7:e7:7e:09:44:cf:0f:5e:b3  .%6.M....~.D..^.
000001c0  5b:47:67:ad:3e:1b:00:a4:0a:9f:e6:d2:e7:e0:65:af  [Gg.>.........e.
000001d0  38:73:c3:72:de:e9:39:6c:a6:2f:51:1a:d9:b9:64:85  8s.r..9l./Q...d.
000001e0  99:b5:e7:83:d8:94:d8:14:b7:e4:0f:df:d2:80:7c:ef  ..............|.
000001f0  b4:1f:2c:62:f3:64:a5:a2:dc:ba:20:94:8c:17:5e:b2  ..,b.d.... ...^.
00000200  a7:34:e4:48:68:43:03:3d:7c:a4:a7:3a:aa:76:0d:40  .4.HhC.=|..:.v.@
00000210  68:95:33:fb:ba:03:66:62:60:4c:6b:d2:6c:4d:0a:00  h.3...fb`Lk.lM..
00000220  00:01:14:00:00:00:0c:72:73:61:2d:73:68:61:32:2d  .......rsa-sha2-
00000230  32:35:36:00:00:01:00:1e:f4:a6:6a:62:01:03:1f:50  256.......jb...P
00000240  cf:8c:dd:df:89:6c:2b:e1:3f:76:8e:8c:d9:10:db:1b  .....l+.?v......
00000250  b6:74:4b:11:db:d0:e6:03:a4:b0:5f:c5:98:dd:b8:f2  .tK......._.....
00000260  95:b1:da:6c:d6:81:bf:81:59:bc:2d:ee:6f:7b:31:95  ...l....Y.-.o{1.
00000270  5f:b3:19:d7:17:83:f8:6d:71:08:2c:26:43:7c:74:99  _......mq.,&C|t.
00000280  8c:c8:7f:e0:0a:49:66:e8:f4:ce:12:ba:45:a9:29:8e  .....If.....E.).
00000290  cb:eb:8e:24:d4:25:7d:10:be:f8:4f:a9:e1:a6:32:15  ...$.%}...O...2.
000002a0  bf:64:fa:3f:38:4c:46:0b:ba:d3:59:15:79:13:3c:79  .d.?8LF...Y.y..y
000002b0  8b:e2:05:bd:04:f8:eb:21:d6:87:b3:e8:d7:a9:89:e0  .......!........
000002c0  11:a0:14:87:27:bb:50:b4:7d:30:ed:af:5c:6d:28:50  ....'.P.}0..\m(P
000002d0  b2:ba:3f:b8:13:3f:f7:0c:27:93:99:65:66:3b:20:73  ..?..?..'..ef; s
000002e0  89:7c:83:8b:ac:33:97:53:55:0e:97:71:21:23:aa:10  .|...3.SU..q!#..
000002f0  be:95:1a:a2:00:c5:df:61:a4:54:40:c7:bb:a9:15:cd  .......a.T@.....
00000300  87:8e:66:50:11:a0:90:0d:af:d6:3c:23:a0:88:13:58  ..fP.......#...X
00000310  e4:e0:51:82:c8:2a:f5:ba:ae:d7:71:d8:99:c4:af:79  ..Q..*....q....y
00000320  15:20:cc:8b:c0:34:0c:20:ac:ba:47:3b:b3:dc:b8:91  . ...4. ..G;....
00000330  5f:f3:45:2d:b7:83:0d                             _.E-...
```

Note how, in that response, there's ssh-rsa and then rsa-sha2-256. ssh-rsa does *not* occur twice.

I'll do some more digging.

**Comment by terrafrost** [comment]
Do you think you could give me the IP address / domain name / port of the server you're connecting to?

Since the problem is with the key exchange I wouldn't need an account on the server - just the IP address / domain name / port and that's it.

Thanks!

**Comment by terrafrost** [comment]
https://github.com/phpseclib/phpseclib/commit/b0d63fbfb56222b3b14bc35aa6f6198b3b326d44 should fix this.

One thought I did have was to do a `user_error('Sever and client agreed to use ' . $server_host_key_algorithm . ' but server returned ' . $this->signature_format . ' signature instead');` after the newly added switch. I've decided against this, for the time being, because, altho that'd work in the 1.0 and 2.0 branches, in the master branch, exceptions are used. ie. if you do user_error the program will continue running. If you throw an exception the program will stop. And idk that this should be a fatal error. I mean, maybe it ought to be but idk.

For the time being, however, I think the current approach is sufficient. If rsa-sha2-256 wasn't available phpseclib would just use ssh-rsa, anyway, so we can just pretend they're the same thing as is. If, in the future, phpseclib has API methods added wherein people can pick and chose which algorithms they want to use, then at that point, throwing an exception might be appropriate, but until such time, I think there's no need for that.

I do wonder how well the server would work with publickey authentication where the publickey is an RSA key that's supposed to be using rsa-sha2-256/512 but I'm not able to test that without an account.
